compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs

```diff
@@ -154,12 +154,31 @@ topologicalSort lfPkgs = map toPkg $ topSort $ transposeG graph
     toPkg = (\(pkg, _, _) -> pkg) . fromVertex
 
 data ReplState = ReplState
-  { moduleNames :: [LF.ModuleName]
+  { imports :: [ImportDecl GhcPs]
   , bindings :: [(LPat GhcPs, Type)]
   , lineNumber :: Int
   }
 
-type ReplM a = Repl.HaskelineT (State.StateT ReplState IO) a
+type ReplM = Repl.HaskelineT (State.StateT ReplState IO)
+
+data ReplInput
+  = ReplStatement (Stmt GhcPs (LHsExpr GhcPs))
+  | ReplImport (ImportDecl GhcPs)
+
+parseReplInput :: String -> DynFlags -> Either Error ReplInput
+parseReplInput input dflags = swapEither $ do
```

Comment:
`Either a b` has a  `Semigroup` instance that should do what you want here and avoids the somewhat confusing `swapEither` thing. Alternatively (no pun intended), you can use https://hackage.haskell.org/package/semigroupoids-5.2.2/docs/Data-Functor-Alt.html#t:Alt which is a bit clearer than the `Semigroup` instance imho.

Reply:
Thanks, I was sure there was one, but ended up looking at the wrong [`Alt`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Alt).

Reply:
Done. The only downside of `Alt` vs the swapped `Either` is that we have to attempt statement parsing last to get statement parse errors on invalid inputs. Shouldn't be a big issue though.